src/pipecat/services/aws_nova_sonic/aws.py

```diff
@@ -248,25 +249,45 @@ async def _handle_input_audio_frame(self, frame: InputAudioRawFrame):
-            # figure out a reliable way to determine when we've gotten our last FINAL text block
-            # after the LLM is done talking.
-            #
-            # First I looked at stopReason, but it doesn't seem like the last FINAL text block is
-            # reliably marked END_TURN (sometimes the *first* one is, but not the last...bug?)
-            #
-            # Then I considered schemes where we tally or match up SPECULATIVE text blocks with
-            # FINAL text blocks to know how many or which FINAL blocks to expect, but user
-            # interruptions throw a wrench in these schemes: depending on the exact timing of the
-            # interruption, we should or shouldn't expect some FINAL blocks.
-            await asyncio.sleep(0.25)
-            self._assistant_is_responding = False
-            await self._report_assistant_response_ended()
+    async def _handle_bot_stopped_speaking(self, delay_to_catch_trailing_assistant_text: bool):
+        # Protect against back-to-back BotStoppedSpeaking calls, which I've observed
+        if self._handling_bot_stopped_speaking:
+            return
+        self._handling_bot_stopped_speaking = True
+
+        async def finalize_assistant_response():
+            if self._assistant_is_responding:
+                # Consider the assistant finished with their response (possibly after a short delay,
+                # to allow for any trailing FINAL assistant text block to come in that need to make
+                # it into context).
+                #
+                # TODO: ideally we could base this solely on the LLM output events, but I couldn't
+                # figure out a reliable way to determine when we've gotten our last FINAL text block
+                # after the LLM is done talking.
+                #
+                # First I looked at stopReason, but it doesn't seem like the last FINAL text block
+                # is reliably marked END_TURN (sometimes the *first* one is, but not the last...
+                # bug?)
+                #
+                # Then I considered schemes where we tally or match up SPECULATIVE text blocks with
+                # FINAL text blocks to know how many or which FINAL blocks to expect, but user
+                # interruptions throw a wrench in these schemes: depending on the exact timing of
+                # the interruption, we should or shouldn't expect some FINAL blocks.
+                if delay_to_catch_trailing_assistant_text:
+                    # This delay length is a balancing act between "catching" trailing assistant
+                    # text that is quite delayed but not waiting so long that user text comes in
+                    # first and results in a bit of context message order scrambling.
+                    await asyncio.sleep(1.25)
+                self._assistant_is_responding = False
+                await self._report_assistant_response_ended()
```

Comment:
Would it make sense if we receive a user text, we check if the `self. _assistant_is_responding` is `True` and if it is, invoke `await self._report_assistant_response_ended()` before doing anything else ?

Reply:
Yeah, and then after recording the user text immediately go back to reporting another assistant response started?

I think we have some more upcoming work to do around aggregating user text as well in the context...maybe we can tackle it then.

Hopefully AWS adds streamed assistant text before too long and all of this complexity can go away. But not sure where that is on the roadmap, or if it's on there.